### PR TITLE
Remove CrateCollector and related components

### PR DESCRIPTION
**Pull request by mfussenegger**
For the most part the CrateCollector was an artifact left over prior to
the BatchIterator introduction where everything was push based.

This changes several components to return a BatchIterator and also
changes the projector application pattern a bit: In most cases the
BatchIterators are now wrapped eagerly instead of lazily when the
consumer is triggered.

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    public Projectors(Collection<? extends Projection> projections,
+                      UUID jobId,
+                      RamAccountingContext ramAccountingContext,
+                      ProjectorFactory projectorFactory) {
+        boolean independentScroll = false;
+        this.projectors = new ArrayList<>(projections.size());
+        for (Projection projection : projections) {
+            if (projection.requiredGranularity().ordinal() > projectorFactory.supportedGranularity().ordinal()) {
+                continue;
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java]: isn't this a programming error? should we at least log something/ add an assertion?

File: sql/src/main/java/io/crate/execution/engine/collect/sources/ShardCollectSource.java [hunk 2]
```diff
@@ -299,23 +299,34 @@ public CrateCollector getCollector(CollectPhase phase,
             );
         }
 
-        RowConsumer firstConsumer = ProjectingRowConsumer.create(
-            lastConsumer,
+        Projectors projectors = new Projectors(
             normalizedPhase.projections(),
-            collectPhase.jobId(),
+            normalizedPhase.jobId(),
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/sources/ShardCollectSource.java]: what's the reason for this change?

File: sql/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java [hunk 3]
```diff
@@ -112,55 +112,33 @@
...
-            builder = RowsCollector.emptyBuilder();
-        } else {
-            assert normalizedCollectNode.maxRowGranularity() == RowGranularity.DOC : "granularity must be DOC";
-            builder = getBuilder(normalizedCollectNode, requiresScroll, collectTask);
-        }
+    public BatchIterator<Row> getIterator(RoutedCollectPhase collectPhase,
+                                          boolean requiresScroll,
+                                          CollectTask collectTask) throws Exception {
+        assert collectPhase.orderBy() == null
+            : "getDocCollector shouldn't be called if there is an orderBy on the collectPhase";
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java]: Is this still relevant? (if so it should say getIterator)

Comment (andreidan) [reply sql/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java]: hmm ... given we have a "getUnorderedIterator" one might assume an orderby here is totally fine no?

File: sql/src/main/java/io/crate/execution/engine/collect/sources/ShardCollectSource.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/sources/ShardCollectSource.java]: No real reason other that my OCD didn't like having `normalizedPhase` followed by `collectPhase`, given that both have the same `jobId`

File: sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java]: Nope, see https://github.com/crate/crate/pull/7480 - I'll add some comments about it

File: sql/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java (hunk 3 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java]: It's still relevant. Before we had `getCollectorBuilder` and `getBuilder`, both with the same signature. Where one is public and one is internal/for the subclass to override. Not sure how to improve the naming. There is also a public `getOrderedCollector` which is why I went with `getUnorderedIterator` for the protected method.

File: sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/pipeline/Projectors.java]: Pushed https://github.com/crate/crate/pull/7505/commits/8b68a8434dabd1aa970263828bb8a9e8eb0fb7d8

File: sql/src/main/java/io/crate/execution/engine/collect/MapSideDataCollectOperation.java [hunk 4]
```diff
@@ -49,36 +50,19 @@ public MapSideDataCollectOperation(CollectSourceResolver collectSourceResolver,
...
-     * &nbsp; -&gt; run shard or doc level collect<br>
-     * * else if we got cluster RowGranularity:<br>
-     * &nbsp; -&gt; run node level collect (cluster level)<br>
-     * </p>
-     */
-    public CrateCollector createCollector(CollectPhase collectPhase,
-                                          RowConsumer consumer,
-                                          final CollectTask collectTask) {
+    public BatchIterator<Row> createIterator(CollectPhase collectPhase,
+                                             RowConsumer consumer,
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/MapSideDataCollectOperation.java]: looks like only `requiresScroll` of the consumer is used here now, so we can just pass this?
